src/coreclr/vm/jitinterface.cpp

```diff
@@ -6817,7 +6817,13 @@ void CEEInfo::setMethodAttribs (
             if (attribs & CORINFO_FLG_SWITCHED_TO_MIN_OPT)
             {
                 _ASSERTE(!ftn->IsJitOptimizationDisabled());
-                config->SetJitSwitchedToMinOpt();
+                // UnmanagedCallersOnly methods that are switched to optimized must be optimized.
```

Comment:
I don't get this code path. From this point it appears the method in question has been marked as `CORINFO_FLG_SWITCHED_TO_MIN_OPT`, but the check in here validates it wasn't "switched to optimized". This begs the question why would the config be "switched to optimized" with the "switched to min opt" flag being passed in? This seems like a fundamental logical flaw. The next question would be below when the flag `CORINFO_FLG_SWITCHED_TO_OPTIMIZED` is passed in - is it okay for that? Even though this method state should be return `false` with `IsEligibleForTieredCompilation`? It may not be but the point here is that methods marked with `UnmanagedCallersOnlyAttribute` aren't supported. Which means now we have a case where `IsEligibleForTieredCompilation` isn't always correct.

Reply:
This path would be taken when the JIT decides to optimize or not, differently from the JIT flags that were requested. It may switch to minopt (for a JIT invocation requesting optimizations) if the method is too complex to optimize, or it may switch to optimized (for a JIT invocation requesting tier 0 JIT) if the method contains a loop since we currently don't tier methods that contain a loop when the method does not have prejitted code. The change here in the `PrepareCodeConfig` is only meant to convey the change up to the invoker of the JIT so that the method would be treated appropriately, for instance to disable tiering for the method if a tier 0 JIT invocation generated optimized code. In that case the method is technically eligible for tiering according to `IsEligibleForTieredCompilation` (which is determined much earlier with less info) but tiering is disabled shortly after jitting it for the first time by disabling call counting. For a method that contains a loop, if it has prejitted code, the method would be tiered, we would just not JIT it without optimizations when it doesn't have prejitted code.

> There's already some code in jithelpers.cpp to force jitted UnmanagedCallersOnly methods to switch to optimized code which works locally in what I thought would be a minimal repro

Come to think of it, that may be what is happening here. Does the method contain a loop?

Since we want to disable tiering for `UnmanagedCallersOnly` methods whether prejitted or not, `MethodDesc::PrepareILBasedCode` may be a good place to disable call counting, using a check similar to that currently done in `MethodDesc::GetPrecompiledCode`. If the method has prejitted code, the prejitted code would get used and call counting would be disabled (same as currently). If the method has to be jitted, the JIT would be invoked with flags requesting optimizations because call counting has already been disabled.

Reply:
No. We explicitly force any method marked with the `CORJIT_REVERSE_PINVOKE` jit flag (which is only applied to UnmanagedCallersOnly-style methods and IL stubs to switch to optimized code in an attempt to opt-out of TC.

Reply:
Oh I see, I guess it was setting the tier1 flag but the JIT then decides to switch to minopt for some reason, in which case tiering would is not disabled